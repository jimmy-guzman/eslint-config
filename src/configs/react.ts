import jsxA11y from "eslint-plugin-jsx-a11y";
import react from "eslint-plugin-react";
import reactHooks from "eslint-plugin-react-hooks";
import reactRefresh from "eslint-plugin-react-refresh";
import globals from "globals";

import { GLOB_JSX, GLOB_TSX } from "../constants";
import { reactRules } from "../rules/react";

const reactConfig = () => {
  return [
    {
      name: "jimmy.codes/react",
      files: [GLOB_JSX, GLOB_TSX],
      plugins: {
        react,
        "react-hooks": reactHooks,
        "jsx-a11y": {
          rules: jsxA11y.rules,
        },
        "react-refresh": reactRefresh,
      },
      languageOptions: {
        parserOptions: {
          jsxPragma: null,
          ecmaFeatures: {
            jsx: true,
          },
        },
        globals: {
          ...globals.browser,
        },
      },
      settings: {
        react: {
          version: "detect",
        },
      },
      rules: reactRules,
    },
  ];
};

export default reactConfig;
