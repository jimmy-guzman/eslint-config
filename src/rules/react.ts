import jsxA11y from "eslint-plugin-jsx-a11y";
import reactPlugin from "eslint-plugin-react";

import type { Rules } from "../types";

import { normalizeRuleEntries } from "../utils/normalize-rule-entries";

export const reactRules = {
  ...jsxA11y.configs.recommended.rules,
  ...normalizeRuleEntries(reactPlugin.configs.flat?.recommended?.rules),
  ...normalizeRuleEntries(reactPlugin.configs.flat?.["jsx-runtime"]?.rules),
  "react/boolean-prop-naming": "off", // revisit
  "react/button-has-type": "error",
  "react/checked-requires-onchange-or-readonly": "error",
  "react/default-props-match-prop-types": "error",
  "react/destructuring-assignment": "off", // revisit
  "react/forbid-component-props": "off",
  "react/forbid-dom-props": "off",
  "react/forbid-elements": "off",
  "react/forbid-foreign-prop-types": "off",
  "react/forbid-prop-types": "off",
  "react/forward-ref-uses-ref": "error",
  "react/function-component-definition": "off", // revisit
  "react/hook-use-state": "error",
  "react/iframe-missing-sandbox": "error",
  "react/jsx-boolean-value": ["error", "never"],
  "react/jsx-curly-brace-presence": "error",
  "react/jsx-filename-extension": "off",
  "react/jsx-fragments": ["error", "syntax"],
  "react/jsx-handler-names": "off",
  "react/jsx-max-depth": "off",
  "react/jsx-no-bind": "off", // revisit
  "react/jsx-no-constructed-context-values": "error",
  "react/jsx-no-leaked-render": "error",
  "react/jsx-no-literals": "off",
  "react/jsx-no-script-url": "error",
  "react/jsx-no-useless-fragment": "error",
  "react/jsx-one-expression-per-line": "off",
  "react/jsx-pascal-case": ["error", { allowNamespace: true }],
  "react/jsx-props-no-spread-multi": "off",
  "react/jsx-props-no-spreading": "off",
  "react/jsx-sort-default-props": "off",
  "react/jsx-sort-props": "off",
  "react/no-access-state-in-setstate": "error",
  "react/no-adjacent-inline-elements": "off",
  "react/no-array-index-key": "off",
  "react/no-arrow-function-lifecycle": "error",
  "react/no-danger": "off",
  "react/no-did-mount-set-state": "error",
  "react/no-did-update-set-state": "error",
  "react/no-invalid-html-attribute": "error",
  "react/no-multi-comp": "off",
  "react/no-namespace": "error",
  "react/no-object-type-as-default-prop": "error",
  "react/no-redundant-should-component-update": "error",
  "react/no-set-state": "off",
  "react/no-this-in-sfc": "error",
  "react/no-typos": "error",
  "react/no-unstable-nested-components": "error",
  "react/no-unused-class-component-methods": "error",
  "react/no-unused-prop-types": "error",
  "react/no-unused-state": "error",
  "react/no-will-update-set-state": "error",
  "react/prefer-es6-class": "off",
  "react/prefer-exact-props": "off",
  "react/prefer-read-only-props": "off",
  "react/prefer-stateless-function": "off",
  "react/require-default-props": "off",
  "react/require-optimization": "off",
  "react/self-closing-comp": "error",
  "react/sort-comp": "off",
  "react/sort-default-props": "off",
  "react/sort-prop-types": "off",
  "react/state-in-constructor": "off",
  "react/static-property-placement": "off",
  "react/style-prop-object": "error",
  "react/void-dom-elements-no-children": "error",
  "react-hooks/exhaustive-deps": "error",
  "react-hooks/rules-of-hooks": "error",
  "react-refresh/only-export-components": [
    "warn",
    { allowConstantExport: true },
  ],
} satisfies Rules;
