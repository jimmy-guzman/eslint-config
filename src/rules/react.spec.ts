import { isPackageExists } from "local-pkg";

import { reactRules } from "./react";

vi.mock("local-pkg");

describe("reactRules", () => {
  it("should create react rules", async () => {
    await expect(reactRules()).resolves.toMatchSnapshot();
  });

  it("should add allowExportNames for nextjs", async () => {
    vi.mocked(isPackageExists).mockImplementation((name) => {
      return name === "next";
    });

    const rules = await reactRules();

    const { allowExportNames } =
      rules["react-refresh/only-export-components"][1];

    expect(allowExportNames).toMatchInlineSnapshot(`
      [
        "dynamic",
        "dynamicParams",
        "revalidate",
        "fetchCache",
        "runtime",
        "preferredRegion",
        "maxDuration",
        "config",
        "generateStaticParams",
        "metadata",
        "generateMetadata",
        "viewport",
        "generateViewport",
      ]
    `);
  });

  it("should only enable allowConstantExport for vite", async () => {
    vi.mocked(isPackageExists).mockImplementation((name) => {
      return name === "vite";
    });

    const rules = await reactRules();

    const { allowConstantExport } =
      rules["react-refresh/only-export-components"][1];

    expect(allowConstantExport).toBe(true);
  });

  it("should enable recommended TypeScript rules", async () => {
    vi.mocked(isPackageExists).mockImplementation((name) => {
      return name === "typescript";
    });

    await expect(reactRules()).resolves.toMatchSnapshot();
  });
});
