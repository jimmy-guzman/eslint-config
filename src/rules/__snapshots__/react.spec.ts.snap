// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`reactRules > should create react rules 1`] = `
{
  "jsx-a11y/alt-text": "error",
  "jsx-a11y/anchor-ambiguous-text": "off",
  "jsx-a11y/anchor-has-content": "error",
  "jsx-a11y/anchor-is-valid": "error",
  "jsx-a11y/aria-activedescendant-has-tabindex": "error",
  "jsx-a11y/aria-props": "error",
  "jsx-a11y/aria-proptypes": "error",
  "jsx-a11y/aria-role": "error",
  "jsx-a11y/aria-unsupported-elements": "error",
  "jsx-a11y/autocomplete-valid": "error",
  "jsx-a11y/click-events-have-key-events": "error",
  "jsx-a11y/control-has-associated-label": [
    "off",
    {
      "ignoreElements": [
        "audio",
        "canvas",
        "embed",
        "input",
        "textarea",
        "tr",
        "video",
      ],
      "ignoreRoles": [
        "grid",
        "listbox",
        "menu",
        "menubar",
        "radiogroup",
        "row",
        "tablist",
        "toolbar",
        "tree",
        "treegrid",
      ],
      "includeRoles": [
        "alert",
        "dialog",
      ],
    },
  ],
  "jsx-a11y/heading-has-content": "error",
  "jsx-a11y/html-has-lang": "error",
  "jsx-a11y/iframe-has-title": "error",
  "jsx-a11y/img-redundant-alt": "error",
  "jsx-a11y/interactive-supports-focus": [
    "error",
    {
      "tabbable": [
        "button",
        "checkbox",
        "link",
        "searchbox",
        "spinbutton",
        "switch",
        "textbox",
      ],
    },
  ],
  "jsx-a11y/label-has-associated-control": "error",
  "jsx-a11y/label-has-for": "off",
  "jsx-a11y/media-has-caption": "error",
  "jsx-a11y/mouse-events-have-key-events": "error",
  "jsx-a11y/no-access-key": "error",
  "jsx-a11y/no-autofocus": "error",
  "jsx-a11y/no-distracting-elements": "error",
  "jsx-a11y/no-interactive-element-to-noninteractive-role": [
    "error",
    {
      "canvas": [
        "img",
      ],
      "tr": [
        "none",
        "presentation",
      ],
    },
  ],
  "jsx-a11y/no-noninteractive-element-interactions": [
    "error",
    {
      "alert": [
        "onKeyUp",
        "onKeyDown",
        "onKeyPress",
      ],
      "body": [
        "onError",
        "onLoad",
      ],
      "dialog": [
        "onKeyUp",
        "onKeyDown",
        "onKeyPress",
      ],
      "handlers": [
        "onClick",
        "onError",
        "onLoad",
        "onMouseDown",
        "onMouseUp",
        "onKeyPress",
        "onKeyDown",
        "onKeyUp",
      ],
      "iframe": [
        "onError",
        "onLoad",
      ],
      "img": [
        "onError",
        "onLoad",
      ],
    },
  ],
  "jsx-a11y/no-noninteractive-element-to-interactive-role": [
    "error",
    {
      "fieldset": [
        "radiogroup",
        "presentation",
      ],
      "li": [
        "menuitem",
        "menuitemradio",
        "menuitemcheckbox",
        "option",
        "row",
        "tab",
        "treeitem",
      ],
      "ol": [
        "listbox",
        "menu",
        "menubar",
        "radiogroup",
        "tablist",
        "tree",
        "treegrid",
      ],
      "table": [
        "grid",
      ],
      "td": [
        "gridcell",
      ],
      "ul": [
        "listbox",
        "menu",
        "menubar",
        "radiogroup",
        "tablist",
        "tree",
        "treegrid",
      ],
    },
  ],
  "jsx-a11y/no-noninteractive-tabindex": [
    "error",
    {
      "allowExpressionValues": true,
      "roles": [
        "tabpanel",
      ],
      "tags": [],
    },
  ],
  "jsx-a11y/no-redundant-roles": "error",
  "jsx-a11y/no-static-element-interactions": [
    "error",
    {
      "allowExpressionValues": true,
      "handlers": [
        "onClick",
        "onMouseDown",
        "onMouseUp",
        "onKeyPress",
        "onKeyDown",
        "onKeyUp",
      ],
    },
  ],
  "jsx-a11y/role-has-required-aria-props": "error",
  "jsx-a11y/role-supports-aria-props": "error",
  "jsx-a11y/scope": "error",
  "jsx-a11y/tabindex-no-positive": "error",
  "react-compiler/react-compiler": "error",
  "react-hooks/exhaustive-deps": "error",
  "react-hooks/rules-of-hooks": "error",
  "react-refresh/only-export-components": [
    "warn",
    {
      "allowConstantExport": undefined,
      "allowExportNames": [],
    },
  ],
  "react/boolean-prop-naming": "off",
  "react/button-has-type": "error",
  "react/checked-requires-onchange-or-readonly": "error",
  "react/default-props-match-prop-types": "error",
  "react/destructuring-assignment": "off",
  "react/display-name": "error",
  "react/forbid-component-props": "off",
  "react/forbid-dom-props": "off",
  "react/forbid-elements": "off",
  "react/forbid-foreign-prop-types": "off",
  "react/forbid-prop-types": "off",
  "react/forward-ref-uses-ref": "error",
  "react/function-component-definition": "off",
  "react/hook-use-state": "error",
  "react/iframe-missing-sandbox": "error",
  "react/jsx-boolean-value": [
    "error",
    "never",
  ],
  "react/jsx-curly-brace-presence": "error",
  "react/jsx-filename-extension": "off",
  "react/jsx-fragments": [
    "error",
    "syntax",
  ],
  "react/jsx-handler-names": "off",
  "react/jsx-key": "error",
  "react/jsx-max-depth": "off",
  "react/jsx-no-bind": "off",
  "react/jsx-no-comment-textnodes": "error",
  "react/jsx-no-constructed-context-values": "error",
  "react/jsx-no-duplicate-props": "error",
  "react/jsx-no-leaked-render": "error",
  "react/jsx-no-literals": "off",
  "react/jsx-no-script-url": "error",
  "react/jsx-no-target-blank": "error",
  "react/jsx-no-undef": "error",
  "react/jsx-no-useless-fragment": "error",
  "react/jsx-one-expression-per-line": "off",
  "react/jsx-pascal-case": [
    "error",
    {
      "allowNamespace": true,
    },
  ],
  "react/jsx-props-no-spread-multi": "off",
  "react/jsx-props-no-spreading": "off",
  "react/jsx-sort-default-props": "off",
  "react/jsx-sort-props": "off",
  "react/jsx-uses-react": "off",
  "react/jsx-uses-vars": "error",
  "react/no-access-state-in-setstate": "error",
  "react/no-adjacent-inline-elements": "off",
  "react/no-array-index-key": "off",
  "react/no-arrow-function-lifecycle": "error",
  "react/no-children-prop": "error",
  "react/no-danger": "off",
  "react/no-danger-with-children": "error",
  "react/no-deprecated": "error",
  "react/no-did-mount-set-state": "error",
  "react/no-did-update-set-state": "error",
  "react/no-direct-mutation-state": "error",
  "react/no-find-dom-node": "error",
  "react/no-invalid-html-attribute": "error",
  "react/no-is-mounted": "error",
  "react/no-multi-comp": "off",
  "react/no-namespace": "error",
  "react/no-object-type-as-default-prop": "error",
  "react/no-redundant-should-component-update": "error",
  "react/no-render-return-value": "error",
  "react/no-set-state": "off",
  "react/no-string-refs": "error",
  "react/no-this-in-sfc": "error",
  "react/no-typos": "error",
  "react/no-unescaped-entities": "error",
  "react/no-unknown-property": "error",
  "react/no-unsafe": "off",
  "react/no-unstable-nested-components": "error",
  "react/no-unused-class-component-methods": "error",
  "react/no-unused-prop-types": "error",
  "react/no-unused-state": "error",
  "react/no-will-update-set-state": "error",
  "react/prefer-es6-class": "off",
  "react/prefer-exact-props": "off",
  "react/prefer-read-only-props": "off",
  "react/prefer-stateless-function": "off",
  "react/prop-types": "error",
  "react/react-in-jsx-scope": "off",
  "react/require-default-props": "off",
  "react/require-optimization": "off",
  "react/require-render-return": "error",
  "react/self-closing-comp": "error",
  "react/sort-comp": "off",
  "react/sort-default-props": "off",
  "react/sort-prop-types": "off",
  "react/state-in-constructor": "off",
  "react/static-property-placement": "off",
  "react/style-prop-object": "error",
  "react/void-dom-elements-no-children": "error",
}
`;
